### PR TITLE
hide shared files located in group folder's trash bin

### LINKED ISSUE #12281: deleted files reappear over and over
### Steps to reproduce
We have a shared Folder, where we deleted files and directories. Not all clients are online at the same time.
In irregular intervals earlier deleted files and directories reappear over and over. We try to delete them, but don´t get rid of them.

### Expected behaviour
Sync the local folder to the server state.
When ready: sync bidirectional.
You can not get rid of files or folders as long as at least one client which collaborate is offline and keep the deleted files. They get pushed back over and over and soon the offline clients get connected.

### Actual behaviour
Sync bidirectional, pushing up all locally kept files/folders, which are not on the server back to it.

## Server configuration detail

**Operating system:** Linux 4.15.0-1025-aws #25-Ubuntu SMP Wed Oct 10 14:23:49 UTC 2018 x86_64

**Webserver:** Apache/2.4.29 (Ubuntu) (apache2handler)

**Database:** mysql 5.6.10

**PHP version:** 

7.0.32-3+ubuntu18.04.1+deb.sury.org+1
Modules loaded: Core, date, libxml, openssl, pcre, zlib, filter, hash, Reflection, SPL, session, standard, apache2handler, mysqlnd, PDO, xml, apcu, apc, calendar, ctype, curl, dom, mbstring, fileinfo, ftp, gd, gettext, iconv, igbinary, imagick, intl, json, ldap, exif, mcrypt, mysqli, pdo_mysql, Phar, posix, readline, redis, shmop, SimpleXML, smbclient, soap, sockets, sysvmsg, sysvsem, sysvshm, tokenizer, wddx, xmlreader, xmlrpc, xmlwriter, xsl, zip, libsmbclient, Zend OPcache

**Nextcloud version:** 14.0.3 - 14.0.3.0

**Updated from an older Nextcloud/ownCloud or fresh install:** 

**Where did you install Nextcloud from:** unknown

<details><summary>Signing status</summary>

Array
(
)

</details>

<details><summary>List of activated apps</summary>

```
Enabled:
 - accessibility: 1.0.1
 - activity: 2.7.0
 - admin_audit: 1.4.0
 - bruteforcesettings: 1.1.0
 - cloud_federation_api: 0.0.1
 - comments: 1.4.0
 - dav: 1.6.0
 - dropit: 0.1.1
 - external: 3.1.0
 - federatedfilesharing: 1.4.0
 - files: 1.9.0
 - files_accesscontrol: 1.4.0
 - files_pdfviewer: 1.3.2
 - files_rightclick: 0.8.4
 - files_sharing: 1.6.2
 - files_texteditor: 2.6.0
 - files_trashbin: 1.4.1
 - files_versions: 1.7.1
 - files_videoplayer: 1.3.0
 - gallery: 18.1.0
 - groupfolders: 1.3.3
 - impersonate: 1.1.0
 - issuetemplate: 0.4.0
 - logreader: 2.0.0
 - lookup_server_connector: 1.2.0
 - metadata: 0.7.0
 - notifications: 2.2.1
 - oauth2: 1.2.1
 - password_policy: 1.4.0
 - previewgenerator: 1.1.0
 - provisioning_api: 1.4.0
 - serverinfo: 1.4.0
 - sharebymail: 1.4.0
 - sharerenamer: 2.2.0
 - socialsharing_email: 1.0.4
 - survey_client: 1.2.0
 - systemtags: 1.4.0
 - theming: 1.5.0
 - twofactor_backupcodes: 1.3.1
 - updatenotification: 1.4.1
 - user_usage_report: 1.1.0
 - w2g2: 2.2.3
 - workflowengine: 1.4.0
Disabled:
 - encryption
 - federation
 - files_downloadactivity
 - files_external
 - files_trackdownloads
 - firstrunwizard
 - nextcloud_announcements
 - support
 - user_external
 - user_ldap

```
</details>

<details><summary>Configuration (config/config.php)</summary>

```
{
    "instanceid": "***REMOVED SENSITIVE VALUE***",
    "passwordsalt": "***REMOVED SENSITIVE VALUE***",
    "secret": "***REMOVED SENSITIVE VALUE***",
    "trusted_domains": [
        "drive.company.com"
    ],
    "datadirectory": "***REMOVED SENSITIVE VALUE***",
    "dbtype": "mysql",
    "version": "14.0.3.0",
    "overwrite.cli.url": "https:\/\/drive.company.com\/",
    "htaccess.RewriteBase": "\/",
    "dbname": "***REMOVED SENSITIVE VALUE***",
    "dbhost": "***REMOVED SENSITIVE VALUE***",
    "dbport": "",
    "dbtableprefix": "oc_",
    "dbuser": "***REMOVED SENSITIVE VALUE***",
    "dbpassword": "***REMOVED SENSITIVE VALUE***",
    "installed": true,
    "logtimezone": "America\/New_York",
    "logdateformat": "l, F d, Y h:i:s a",
    "log_rotate_size": 10485760,
    "logfile": "\/var\/log\/nextcloud\/nextcloud.log",
    "loglevel": 3,
    "cron_log": true,
    "syslog_tag": "Nextcloud",
    "mail_from_address": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpmode": "smtp",
    "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpport": 587,
    "mail_smtptimeout": 10,
    "mail_smtpsecure": "tls",
    "mail_smtpauth": true,
    "mail_smtpauthtype": "LOGIN",
    "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
    "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
    "mail_domain": "***REMOVED SENSITIVE VALUE***",
    "maintenance": false,
    "debug": false,
    "wnd.logging.enable": false,
    "filelocking.debug": false,
    "memcache.local": "\\OC\\Memcache\\APCu",
    "filelocking.enabled": "true",
    "memcache.locking": "\\OC\\Memcache\\Redis",
    "redis": {
        "host": "***REMOVED SENSITIVE VALUE***",
        "port": 6379,
        "timeout": 0,
        "password": "***REMOVED SENSITIVE VALUE***"
    },
    "enable_avatars": true,
    "knowledgebaseenabled": false,
    "remember_login_cookie_lifetime": 1296000,
    "trashbin_retention_obligation": "30, 31",
    "versions_retention_obligation": "30, 31",
    "enable_previews": true,
    "updater.server.url": "https:\/\/updates.nextcloud.com\/updater_server\/",
    "updater.release.channel": "stable",
    "check_for_working_webdav": true,
    "check_for_working_wellknown_setup": true,
    "has_internet_connection": true,
    "check_for_working_htaccess": true,
    "cipher": "AES-256-CFB",
    "minimum.supported.desktop.version": "1.7.0",
    "singleuser": false,
    "theme": "",
    "appcodechecker": true,
    "activity_expire_days": 365,
    "data-fingerprint": "8b40196ad1c7fb6bc7c6512fb8ae9f79"
}
```
</details>

**Are you using external storage, if yes which one:** local/smb/sftp/...

**Are you using encryption:** 

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...

## Client configuration

**Browser:** Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:63.0) Gecko/20100101 Firefox/63.0

**Operating system:** Windows 10 Pro

### PR DESCRIPTION
**Pull request by blizzz**
* Resolves: #12281

## Summary

- Shares from a group folder appear in the files list even if they are deleted
- the code checking whether a file is accessible was laid out only for the regular trash bin
- Needs one manual backport, because `str_starts_with` is not available in PHP 7.4

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] ~~Screenshots before/after for front-end changes~~
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: lib/private/Share20/DefaultShareProvider.php [hunk 1]
```diff
@@ -885,6 +886,11 @@ private function isAccessibleResult($data) {
 		if ($pathSections[0] !== 'files'
 			&& (strpos($data['storage_string_id'], 'home::') === 0 || strpos($data['storage_string_id'], 'object::user') === 0)) {
 			return false;
+		} else if ($pathSections[0] === '__groupfolders'
```
Comment (skjnldsv) [inline lib/private/Share20/DefaultShareProvider.php:889]: We don't have a global variable for this `__groupfolders` ?

**Review by skjnldsv — APPROVED** [review]

**Comment by come-nc** [comment]
> Needs one manual backport, because str_starts_with is not available in PHP 7.4

We have a polyfill for it from symfony if I remember correctly, so it should work as well on 7.4.

**Review by come-nc — APPROVED** [review]

**Pull request merged**